File: packages/sdk-python/src/agenttwin/redaction.py
"""Client-side redaction of secrets and personal data.

The rules and their semantics mirror ``gokit/redact`` exactly (Go uses RE2;
patterns here use ASCII classes and Go's ``\\s`` set ``[\\t\\n\\f\\r ]`` so
both engines match the same text). Parity is verified against
``packages/contracts/fixtures/redaction.json`` and by differential tests that
run random inputs through the Go reference implementation.

Strategies:

``mask``  replace a match with ``[REDACTED:<kind>]``
``hash``  replace a match with ``[HASH:<kind>:<12 hex of sha256>]`` so equal
          values stay comparable without being revealed
``drop``  drop the whole value when anything sensitive is found
"""

from __future__ import annotations

import copy
import hashlib
import re
from collections.abc import Callable, Iterable, Mapping, Sequence
from dataclasses import dataclass, field
from typing import Any, Literal

__all__ = [
    "PII_RULES",
    "SECRET_RULES",
    "RedactionConfig",
    "Redactor",
    "Rule",
    "Strategy",
    "truncate",
]

Strategy = Literal["mask", "hash", "drop"]

_WS = r"[\t\n\f\r ]"  # Go RE2 \s


@dataclass(frozen=True)
class Rule:
    """One detector. ``group`` selects the sub-match that is replaced."""

    kind: str
    pattern: re.Pattern[str]
    group: int = 0
    valid: Callable[[str], bool] | None = None


def _luhn(s: str) -> bool:
    digits = [ord(c) - 48 for c in s if "0" <= c <= "9"]
    if not 13 <= len(digits) <= 19:
        return False
    total = 0
    double = False
    for d in reversed(digits):
        if double:
            d *= 2
            if d > 9:
                d -= 9
        total += d
        double = not double
    return total % 10 == 0


def _phone_digits(s: str) -> bool:
    n = sum(1 for c in s if "0" <= c <= "9")
    return 9 <= n <= 15


def _re(pattern: str, flags: int = 0) -> re.Pattern[str]:
    return re.compile(pattern, flags | re.ASCII)


SECRET_RULES: tuple[Rule, ...] = (
    Rule("private_key", _re(r"-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----")),
    Rule("jwt", _re(r"\beyJ[A-Za-z0-9_-]{5,}\.[A-Za-z0-9_-]{5,}\.[A-Za-z0-9_-]{5,}")),
    Rule("bearer", _re(rf"\bbearer{_WS}+([A-Za-z0-9._~+/=-]{{8,}})", re.IGNORECASE), group=1),
    Rule(
        "api_key",
        _re(
            r"\b(?:atk_[a-z0-9]{8}_[A-Za-z0-9_-]{8,}|sk-[A-Za-z0-9_-]{16,}|AKIA[0-9A-Z]{16}"
            r"|gh[pousr]_[A-Za-z0-9]{30,}|xox[baprs]-[A-Za-z0-9-]{10,})"
        ),
    ),
    Rule(
        "credential",
        _re(
            r"\b(?:password|passwd|secret|api[_-]?key|access[_-]?token|client[_-]?secret)\b"
            rf"[\"']?{_WS}*[:=]{_WS}*[\"']?([^\t\n\f\r \"',;]{{4,}})",
            re.IGNORECASE,
        ),
        group=1,
    ),
)

PII_RULES: tuple[Rule, ...] = (
    Rule("email", _re(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b")),
    Rule("card", _re(r"\b(?:\d[ -]?){12,18}\d\b"), valid=_luhn),
    Rule(
        "phone",
        _re(
            rf"\+\d{{1,3}}(?:[\t\n\f\r .-]?\d{{1,4}}){{2,5}}\b"
            rf"|\(\d{{3}}\){_WS}?\d{{3}}[\t\n\f\r .-]\d{{4}}\b"
            r"|\b\d{3}[.-]\d{3}[.-]\d{4}\b"
        ),
        valid=_phone_digits,
    ),
)

_MARKER = _re(r"\[(?:REDACTED:[a-z_]+|HASH:[a-z_]+:[0-9a-f]{12})\]")
_MAX_ROUNDS = 8


@dataclass(frozen=True)
class RedactionConfig:
    """What the SDK redacts before any content leaves the process.

    ``pii`` enables email/card/phone detection (secrets are always detected).
    ``custom_patterns`` are extra regular expressions (kind ``custom``).
    ``json_paths`` redact whole fields of structured values such as tool
    arguments, e.g. ``$.customer.email`` or ``$.items[*].card_number``.
    """

    strategy: Strategy = "mask"
    pii: bool = True
    custom_patterns: tuple[str, ...] = ()
    json_paths: tuple[str, ...] = ()

    def redactor(self) -> Redactor:
        rules = list(SECRET_RULES)
        if self.pii:
            rules.extend(PII_RULES)
        rules.extend(Rule("custom", re.compile(p)) for p in self.custom_patterns)
        return Redactor(rules, self.strategy, json_paths=self.json_paths)


@dataclass
class Redactor:
    """Applies rules with a strategy. Instances are immutable after creation."""

    rules: Sequence[Rule]
    strategy: Strategy = "mask"
    json_paths: Sequence[str] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        self._paths = [_parse_path(p) for p in self.json_paths]

    @classmethod
    def secrets(cls, strategy: Strategy = "mask") -> Redactor:
        return cls(SECRET_RULES, strategy)

    @classmethod
    def all(cls, strategy: Strategy = "mask") -> Redactor:
        return cls((*SECRET_RULES, *PII_RULES), strategy)

    # -- text -------------------------------------------------------------

    def text(self, s: str) -> tuple[str | None, bool]:
        """Redact free text. Returns (result, changed); with the ``drop``
        strategy the result is None when anything sensitive was found."""
        out, changed = self._fixpoint(s)
        if changed and self.strategy == "drop":
            return None, True
        return out, changed

    def _fixpoint(self, s: str) -> tuple[str, bool]:
        changed = False
        for _ in range(_MAX_ROUNDS):
            out, c = self._pass(s)
            if not c:
                break
            s, changed = out, True
        return s, changed

    def _pass(self, s: str) -> tuple[str, bool]:
        changed = False
        for rule in self.rules:
            s, c = self._outside_markers(rule, s)
            changed = changed or c
        return s, changed

    def _outside_markers(self, rule: Rule, s: str) -> tuple[str, bool]:
        spans = [m.span() for m in _MARKER.finditer(s)]
        if not spans:
            return self._apply(rule, s)
        parts: list[str] = []
        changed = False
        prev = 0
        for start, end in spans:
            seg, c = self._apply(rule, s[prev:start])
            parts.append(seg)
            parts.append(s[start:end])
            changed = changed or c
            prev = end
        seg, c = self._apply(rule, s[prev:])
        parts.append(seg)
        return "".join(parts), changed or c

    def _apply(self, rule: Rule, s: str) -> tuple[str, bool]:
        parts: list[str] = []
        prev = 0
        changed = False
        for m in rule.pattern.finditer(s):
            start, end = m.span(rule.group)
            if start < 0:
                continue
            target = s[start:end]
            if not target or (rule.valid is not None and not rule.valid(target)):
                continue
            parts.append(s[prev:start])
            parts.append(self._replacement(rule.kind, target))
            prev = end
            changed = True
        if not changed:
            return s, False
        parts.append(s[prev:])
        return "".join(parts), True

    def _replacement(self, kind: str, value: str) -> str:
        if self.strategy == "hash":
            digest = hashlib.sha256(value.encode("utf-8")).hexdigest()[:12]
            return f"[HASH:{kind}:{digest}]"
        return f"[REDACTED:{kind}]"

    # -- structured values ------------------------------------------------

    def value(self, v: Any) -> tuple[Any, bool]:
        """Redact a JSON-like value: configured JSON paths first (whole
        fields), then every string leaf. Keys are never redacted. With the
        ``drop`` strategy, sensitive fields are removed."""
        out = copy.deepcopy(v)
        changed = False
        for path in self._paths:
            out, c = _apply_path(out, path, self)
            changed = changed or c
        out, c = self._leaves(out)
        return out, changed or c

    def _leaves(self, v: Any) -> tuple[Any, bool]:
        if isinstance(v, str):
            red, changed = self.text(v)
            return red, changed
        if isinstance(v, Mapping):
            result: dict[str, Any] = {}
            changed = False
            for k, item in v.items():
                red, c = self._leaves(item)
                changed = changed or c
                if red is None and c and self.strategy == "drop":
                    continue
                result[k] = red
            return result, changed
        if isinstance(v, list | tuple):
            items: list[Any] = []
            changed = False
            for item in v:
                red, c = self._leaves(item)
                changed = changed or c
                if red is None and c and self.strategy == "drop":
                    continue
                items.append(red)
            return items, changed
        return v, False

    def field_replacement(self, v: Any) -> str | None:
        if self.strategy == "drop":
            return None
        if self.strategy == "hash":
            from agenttwin.hashing import canonical_json

            try:
                encoded = canonical_json(v)
            except (TypeError, ValueError):
                encoded = repr(v)
            return f"[HASH:field:{hashlib.sha256(encoded.encode('utf-8')).hexdigest()[:12]}]"
        return "[REDACTED:field]"


# -- JSON paths (a deliberately small subset: $.a.b, $.a[*].b, $.a[0], $['k']) --

_Segment = str | int | Literal["*"]
_PATH_TOKEN = re.compile(r"\.([A-Za-z0-9_\-]+)|\[(\*|\d+)\]|\['([^']*)'\]")


def _parse_path(path: str) -> list[_Segment]:
    if not path.startswith("$"):
        raise ValueError(f"json path must start with '$': {path!r}")
    segs: list[_Segment] = []
    pos = 1
    while pos < len(path):
        m = _PATH_TOKEN.match(path, pos)
        if not m:
            raise ValueError(f"unsupported json path syntax at {pos}: {path!r}")
        name, index, quoted = m.groups()
        if name is not None:
            segs.append(name)
        elif quoted is not None:
            segs.append(quoted)
        elif index == "*":
            segs.append("*")
        else:
            segs.append(int(index))
        pos = m.end()
    if not segs:
        raise ValueError(f"json path selects the whole document: {path!r}")
    return segs


def _apply_path(v: Any, path: list[_Segment], red: Redactor) -> tuple[Any, bool]:
    if not path:
        return v, False
    head, rest = path[0], path[1:]
    changed = False
    if isinstance(v, dict):
        keys: Iterable[str] = list(v.keys()) if head == "*" else ([head] if isinstance(head, str) else [])
        for k in keys:
            if k not in v:
                continue
            if rest:
                v[k], c = _apply_path(v[k], rest, red)
            else:
                rep = red.field_replacement(v[k])
                if rep is None:
                    del v[k]
                else:
                    v[k] = rep
                c = True
            changed = changed or c
    elif isinstance(v, list):
        if head == "*":
            idxs = list(range(len(v)))
        elif isinstance(head, int) and head < len(v):
            idxs = [head]
        else:
            idxs = []
        drop: list[int] = []
        for i in idxs:
            if rest:
                v[i], c = _apply_path(v[i], rest, red)
            else:
                rep = red.field_replacement(v[i])
                if rep is None:
                    drop.append(i)
                else:
                    v[i] = rep
                c = True
            changed = changed or c
        for i in reversed(drop):
            del v[i]
    return v, changed


def truncate(s: str, max_bytes: int) -> tuple[str, bool]:
    """Shorten ``s`` to at most ``max_bytes`` UTF-8 bytes on a character
    boundary, appending a marker (same contract as gokit/redact.Truncate)."""
    raw = s.encode("utf-8")
    if max_bytes <= 0 or len(raw) <= max_bytes:
        return s, False
    marker = "…[truncated]"
    cut = max(max_bytes - len(marker.encode("utf-8")), 0)
    while cut > 0 and (raw[cut] & 0xC0) == 0x80:
        cut -= 1
    return raw[:cut].decode("utf-8") + marker, True

File: packages/sdk-typescript/src/redaction.ts
/**
 * Client-side redaction of secrets and personal data.
 *
 * The rules and their semantics mirror `gokit/redact` and the Python SDK
 * exactly. The patterns use ASCII classes, Go's `\s` set `[\t\n\f\r ]` and no
 * `u` flag, so JavaScript, RE2 and Python's `re` match the same text; parity
 * is verified against `packages/contracts/fixtures/redaction.json`.
 *
 * Strategies:
 *
 * - `mask`: replace a match with `[REDACTED:<kind>]`
 * - `hash`: replace a match with `[HASH:<kind>:<12 hex of sha256>]`, so equal
 *   values stay comparable without being revealed
 * - `drop`: drop the whole value when anything sensitive is found
 */

import { canonicalJson, sha256Hex } from "./hashing.js";

export type Strategy = "mask" | "hash" | "drop";
export const STRATEGIES: readonly Strategy[] = ["mask", "hash", "drop"];

/** One detector. `group` selects the sub-match that is replaced. */
export interface Rule {
  readonly kind: string;
  readonly pattern: RegExp;
  readonly group?: number;
  readonly valid?: (match: string) => boolean;
}

const WS = "[\\t\\n\\f\\r ]"; // Go RE2 \s

function re(source: string, flags = ""): RegExp {
  return new RegExp(source, `gd${flags}`);
}

function luhn(s: string): boolean {
  const digits = [...s].filter((c) => c >= "0" && c <= "9").map((c) => c.charCodeAt(0) - 48);
  if (digits.length < 13 || digits.length > 19) return false;
  let total = 0;
  let double = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let d = digits[i]!;
    if (double) {
      d *= 2;
      if (d > 9) d -= 9;
    }
    total += d;
    double = !double;
  }
  return total % 10 === 0;
}

function phoneDigits(s: string): boolean {
  const n = [...s].filter((c) => c >= "0" && c <= "9").length;
  return n >= 9 && n <= 15;
}

export const SECRET_RULES: readonly Rule[] = [
  {
    kind: "private_key",
    pattern: re("-----BEGIN [A-Z ]*PRIVATE KEY-----[\\s\\S]*?-----END [A-Z ]*PRIVATE KEY-----"),
  },
  { kind: "jwt", pattern: re("\\beyJ[A-Za-z0-9_-]{5,}\\.[A-Za-z0-9_-]{5,}\\.[A-Za-z0-9_-]{5,}") },
  { kind: "bearer", pattern: re(`\\bbearer${WS}+([A-Za-z0-9._~+/=-]{8,})`, "i"), group: 1 },
  {
    kind: "api_key",
    pattern: re(
      "\\b(?:atk_[a-z0-9]{8}_[A-Za-z0-9_-]{8,}|sk-[A-Za-z0-9_-]{16,}|AKIA[0-9A-Z]{16}" +
        "|gh[pousr]_[A-Za-z0-9]{30,}|xox[baprs]-[A-Za-z0-9-]{10,})",
    ),
  },
  {
    kind: "credential",
    pattern: re(
      "\\b(?:password|passwd|secret|api[_-]?key|access[_-]?token|client[_-]?secret)\\b" +
        `["']?${WS}*[:=]${WS}*["']?([^\\t\\n\\f\\r "',;]{4,})`,
      "i",
    ),
    group: 1,
  },
];

export const PII_RULES: readonly Rule[] = [
  { kind: "email", pattern: re("\\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}\\b") },
  { kind: "card", pattern: re("\\b(?:[0-9][ -]?){12,18}[0-9]\\b"), valid: luhn },
  {
    kind: "phone",
    pattern: re(
      "\\+[0-9]{1,3}(?:[\\t\\n\\f\\r .-]?[0-9]{1,4}){2,5}\\b" +
        `|\\([0-9]{3}\\)${WS}?[0-9]{3}[\\t\\n\\f\\r .-][0-9]{4}\\b` +
        "|\\b[0-9]{3}[.-][0-9]{3}[.-][0-9]{4}\\b",
    ),
    valid: phoneDigits,
  },
];

const MARKER = /\[(?:REDACTED:[a-z_]+|HASH:[a-z_]+:[0-9a-f]{12})\]/g;
const MAX_ROUNDS = 8;

/** What the SDK redacts before any content leaves the process. */
export interface RedactionConfig {
  /** `mask` (default), `hash` or `drop`. */
  readonly strategy?: Strategy;
  /** Email, card and phone detection (secrets are always detected). Default true. */
  readonly pii?: boolean;
  /** Extra regular expressions (JavaScript syntax), reported as kind `custom`. */
  readonly customPatterns?: readonly string[];
  /**
   * Whole fields of structured values (tool arguments, results) to redact,
   * e.g. `$.customer.email` or `$.items[*].card_number`.
   */
  readonly jsonPaths?: readonly string[];
}

/** Custom patterns as rules (kind `custom`); throws SyntaxError on a bad pattern. */
export function customRules(patterns: readonly string[] = []): Rule[] {
  return patterns.map((p) => ({ kind: "custom", pattern: new RegExp(p, "gd") }));
}

/** The redactor a configuration describes. */
export function redactorFor(config: RedactionConfig = {}): Redactor {
  const rules: Rule[] = [...SECRET_RULES];
  if (config.pii ?? true) rules.push(...PII_RULES);
  rules.push(...customRules(config.customPatterns));
  return new Redactor(rules, config.strategy ?? "mask", config.jsonPaths ?? []);
}

type Segment = string | number;

/** Applies rules with a strategy. Instances are immutable. */
export class Redactor {
  readonly rules: readonly Rule[];
  readonly strategy: Strategy;
  private readonly paths: Segment[][];

  constructor(rules: readonly Rule[], strategy: Strategy = "mask", jsonPaths: readonly string[] = []) {
    if (!STRATEGIES.includes(strategy)) throw new RangeError(`unknown redaction strategy ${strategy}`);
    this.rules = rules.map((r) => ({ ...r, pattern: globalWithIndices(r.pattern) }));
    this.strategy = strategy;
    this.paths = jsonPaths.map(parsePath);
  }

  static secrets(strategy: Strategy = "mask"): Redactor {
    return new Redactor(SECRET_RULES, strategy);
  }

  static all(strategy: Strategy = "mask"): Redactor {
    return new Redactor([...SECRET_RULES, ...PII_RULES], strategy);
  }

  /**
   * Redacts free text. With the `drop` strategy the text is `null` when
   * anything sensitive was found.
   */
  text(s: string): { text: string | null; changed: boolean } {
    let changed = false;
    for (let round = 0; round < MAX_ROUNDS; round++) {
      const pass = this.pass(s);
      if (!pass.changed) break;
      s = pass.text;
      changed = true;
    }
    if (changed && this.strategy === "drop") return { text: null, changed };
    return { text: s, changed };
  }

  private pass(s: string): { text: string; changed: boolean } {
    let changed = false;
    for (const rule of this.rules) {
      const r = this.outsideMarkers(rule, s);
      s = r.text;
      changed ||= r.changed;
    }
    return { text: s, changed };
  }

  private outsideMarkers(rule: Rule, s: string): { text: string; changed: boolean } {
    const spans = [...s.matchAll(MARKER)].map((m) => [m.index, m.index + m[0].length] as const);
    if (spans.length === 0) return this.apply(rule, s);
    const parts: string[] = [];
    let changed = false;
    let prev = 0;
    for (const [start, end] of spans) {
      const seg = this.apply(rule, s.slice(prev, start));
      parts.push(seg.text, s.slice(start, end));
      changed ||= seg.changed;
      prev = end;
    }
    const seg = this.apply(rule, s.slice(prev));
    parts.push(seg.text);
    return { text: parts.join(""), changed: changed || seg.changed };
  }

  private apply(rule: Rule, s: string): { text: string; changed: boolean } {
    const parts: string[] = [];
    let prev = 0;
    let changed = false;
    for (const m of s.matchAll(rule.pattern)) {
      const span = m.indices?.[rule.group ?? 0];
      if (span === undefined) continue;
      const [start, end] = span;
      const target = s.slice(start, end);
      if (!target || (rule.valid !== undefined && !rule.valid(target))) continue;
      parts.push(s.slice(prev, start), this.replacement(rule.kind, target));
      prev = end;
      changed = true;
    }
    if (!changed) return { text: s, changed: false };
    parts.push(s.slice(prev));
    return { text: parts.join(""), changed: true };
  }

  private replacement(kind: string, value: string): string {
    if (this.strategy === "hash") return `[HASH:${kind}:${sha256Hex(value).slice(0, 12)}]`;
    return `[REDACTED:${kind}]`;
  }

  /**
   * Redacts a JSON value: configured JSON paths first (whole fields), then
   * every string leaf. Keys are never redacted. With the `drop` strategy
   * sensitive fields are removed. The input is not modified.
   */
  value(v: unknown): { value: unknown; changed: boolean } {
    let out = structuredClone(v);
    let changed = false;
    for (const path of this.paths) {
      const r = applyPath(out, path, this);
      out = r.value;
      changed ||= r.changed;
    }
    const r = this.leaves(out);
    return { value: r.value, changed: changed || r.changed };
  }

  private leaves(v: unknown): { value: unknown; changed: boolean } {
    if (typeof v === "string") {
      const r = this.text(v);
      return { value: r.text, changed: r.changed };
    }
    if (Array.isArray(v)) {
      const items: unknown[] = [];
      let changed = false;
      for (const item of v) {
        const r = this.leaves(item);
        changed ||= r.changed;
        if (r.value === null && r.changed && this.strategy === "drop") continue;
        items.push(r.value);
      }
      return { value: items, changed };
    }
    if (v !== null && typeof v === "object") {
      const result: Record<string, unknown> = {};
      let changed = false;
      for (const [k, item] of Object.entries(v)) {
        const r = this.leaves(item);
        changed ||= r.changed;
        if (r.value === null && r.changed && this.strategy === "drop") continue;
        result[k] = r.value;
      }
      return { value: result, changed };
    }
    return { value: v, changed: false };
  }

  /** What a whole field selected by a JSON path becomes (`undefined`: removed). */
  fieldReplacement(v: unknown): string | undefined {
    if (this.strategy === "drop") return undefined;
    if (this.strategy === "hash") {
      let encoded: string;
      try {
        encoded = canonicalJson(v);
      } catch {
        encoded = String(v);
      }
      return `[HASH:field:${sha256Hex(encoded).slice(0, 12)}]`;
    }
    return "[REDACTED:field]";
  }
}

function globalWithIndices(pattern: RegExp): RegExp {
  let flags = pattern.flags;
  if (!flags.includes("g")) flags += "g";
  if (!flags.includes("d")) flags += "d";
  return flags === pattern.flags ? pattern : new RegExp(pattern.source, flags);
}

// JSON paths: a deliberately small subset ($.a.b, $.a[*].b, $.a[0], $['k']).
const PATH_TOKEN = /\.([A-Za-z0-9_-]+)|\[(\*|[0-9]+)\]|\['([^']*)'\]/y;

export function parsePath(path: string): Segment[] {
  if (!path.startsWith("$")) throw new SyntaxError(`json path must start with '$': ${path}`);
  const segs: Segment[] = [];
  let pos = 1;
  while (pos < path.length) {
    PATH_TOKEN.lastIndex = pos;
    const m = PATH_TOKEN.exec(path);
    if (!m) throw new SyntaxError(`unsupported json path syntax at ${pos}: ${path}`);
    const [, name, index, quoted] = m;
    if (name !== undefined) segs.push(name);
    else if (quoted !== undefined) segs.push(quoted);
    else if (index === "*") segs.push("*");
    else segs.push(Number(index));
    pos = PATH_TOKEN.lastIndex;
  }
  if (segs.length === 0) throw new SyntaxError(`json path selects the whole document: ${path}`);
  return segs;
}

function applyPath(v: unknown, path: Segment[], red: Redactor): { value: unknown; changed: boolean } {
  if (path.length === 0) return { value: v, changed: false };
  const [head, ...rest] = path as [Segment, ...Segment[]];
  let changed = false;
  if (Array.isArray(v)) {
    const idxs =
      head === "*" ? v.map((_, i) => i) : typeof head === "number" && head < v.length ? [head] : [];
    const drop: number[] = [];
    for (const i of idxs) {
      if (rest.length > 0) {
        const r = applyPath(v[i], rest, red);
        v[i] = r.value;
        changed ||= r.changed;
      } else {
        const rep = red.fieldReplacement(v[i]);
        if (rep === undefined) drop.push(i);
        else v[i] = rep;
        changed = true;
      }
    }
    for (const i of drop.reverse()) v.splice(i, 1);
  } else if (v !== null && typeof v === "object") {
    const obj = v as Record<string, unknown>;
    const keys = head === "*" ? Object.keys(obj) : typeof head === "string" ? [head] : [];
    for (const k of keys) {
      if (!Object.hasOwn(obj, k)) continue;
      if (rest.length > 0) {
        const r = applyPath(obj[k], rest, red);
        obj[k] = r.value;
        changed ||= r.changed;
      } else {
        const rep = red.fieldReplacement(obj[k]);
        if (rep === undefined) Reflect.deleteProperty(obj, k);
        else obj[k] = rep;
        changed = true;
      }
    }
  }
  return { value: v, changed };
}

const TRUNCATED = "…[truncated]";
const TRUNCATED_BYTES = Buffer.byteLength(TRUNCATED);

/**
 * Shortens `s` to at most `maxBytes` UTF-8 bytes on a character boundary,
 * appending a marker (same contract as gokit/redact.Truncate).
 */
export function truncate(s: string, maxBytes: number): { text: string; truncated: boolean } {
  const raw = Buffer.from(s, "utf8");
  if (maxBytes <= 0 || raw.length <= maxBytes) return { text: s, truncated: false };
  let cut = Math.max(maxBytes - TRUNCATED_BYTES, 0);
  while (cut > 0 && (raw[cut]! & 0xc0) === 0x80) cut--;
  return { text: raw.subarray(0, cut).toString("utf8") + TRUNCATED, truncated: true };
}
